src/main/java/uk/gov/ch/developer/docs/controller/developer/UserCallbackController.java

```diff
@@ -41,22 +52,90 @@ public String getCallback(@RequestParam("state") String state,
             // Nonce values
         }
 
-        WebClient webClient = WebClient.create();
+        final Session chSession = (Session) httpServletRequest
+                .getAttribute(SessionHandler.CHS_SESSION_REQUEST_ATT_KEY);
 
-        webClient.post().uri(URI.create(identityProvider.getTokenUrl()))
-                .contentType(MediaType.APPLICATION_FORM_URLENCODED)
-                .accept(MediaType.APPLICATION_JSON)
-                .body(BodyInserters.fromPublisher(Mono.just(getPostRequestBody(code)),
-                        String.class))
-                .retrieve().bodyToMono(OAuthToken.class)
-                .subscribe(s -> LOGGER.trace("Access Token: " + s.getAccessToken()));// TODO remove
-                                                                                     // this log
-                                                                                     // line
+        Map<String, Object> signInData = new HashMap<>();
+
+        Mono<OAuthToken> oauthTokenResponse = getOAuthToken(code, webClient);
+
+        LOGGER.debug("Getting User Profile....");
+
+        UserProfileResponse userProfileResponse = getUserProfile(signInData, webClient, oauthTokenResponse);
+
+        if (userProfileResponse != null) {
+
+            LOGGER.debug("Processing Response");
+            setUserProfile(signInData, userProfileResponse);
+            signInData.put(SessionKeys.SIGNED_IN.getKey(), Integer.valueOf(1));
+            chSession.getData().put(SessionKeys.SIGN_IN_INFO.getKey(), signInData);
+
+        }
+
+        LOGGER.debug("Redirecting...");
 
         return ("redirect:" + identityProvider.getRedirectUriPage());// TODO redirect back to page
                                                                      // where sign-in was initiated
     }
 
+    private UserProfileResponse getUserProfile(Map<String, Object> signInData, WebClient webClient,
+            Mono<OAuthToken> oauthTokenResponse) {
+        LOGGER.debug("Getting User Profile....");
+
+        OAuthToken oauthToken = oauthTokenResponse.block();
+
+        return requestUserProfile(signInData, webClient, oauthToken);
+    }
+
+    private UserProfileResponse requestUserProfile(Map<String, Object> signInData, WebClient webClient,
+            OAuthToken tokenResponse) {
+        setAccessToken(signInData, tokenResponse);
```

Comment:
Why are we doing this in this method? I think it would be better that if signInData is not needed in this method I'm worried this might obscure the trace of building up our user data.